compositor_render/src/transformations/web_renderer/chromium_sender/thread_state.rs

```diff
@@ -0,0 +1,98 @@
+use std::{collections::HashMap, path::PathBuf};
+
+use compositor_chromium::cef;
+use compositor_common::scene::NodeId;
+
+use crate::{
+    transformations::web_renderer::{shared_memory::SharedMemory, WebRenderer},
+    UNEMBED_SOURCE_MESSAGE,
+};
+
+use super::chromium_sender_thread::ChromiumSenderThreadError;
+
+pub(super) struct ThreadState {
+    pub browser: cef::Browser,
+    pub shared_memory: HashMap<NodeId, Vec<SharedMemory>>,
+    pub shared_memory_root_path: PathBuf,
+    pub pending_resizes: HashMap<PathBuf, PendingResize>,
+}
+
+impl ThreadState {
+    pub fn new(browser: cef::Browser, renderer_id: &str) -> Self {
+        let shared_memory_root_path = WebRenderer::shared_memory_root_path(renderer_id);
+        let shared_memory = HashMap::new();
+        let pending_resizes = HashMap::new();
+
+        Self {
+            browser,
+            shared_memory,
+            shared_memory_root_path,
+            pending_resizes,
+        }
+    }
+
+    pub fn shared_memory(
+        &self,
+        node_id: &NodeId,
+        source_idx: usize,
+    ) -> Result<&SharedMemory, ChromiumSenderThreadError> {
+        let node_shared_memory = self
+            .shared_memory
+            .get(node_id)
+            .ok_or_else(|| ChromiumSenderThreadError::SharedMemoryNotAllocated(node_id.clone()))?;
+
+        Ok(&node_shared_memory[source_idx])
+    }
+
+    pub fn shared_memory_mut(
+        &mut self,
+        node_id: &NodeId,
+        source_idx: usize,
+    ) -> Result<&mut SharedMemory, ChromiumSenderThreadError> {
+        let node_shared_memory = self
+            .shared_memory
+            .get_mut(node_id)
```

Comment:
Why `mut`?

Reply:
I want to return a mutable reference to `SharedMemory` which is stored in `HashMap<NodeId, Vec<SharedMemory>>`. I have to borrow mutably `Vec<SharedMemory>` first so that I could return a mutable reference to one of the elements